### PR TITLE
Server Facade

### LINKED ISSUE #259: Rename `Server` to `StreamingServer` and add new `Server` facade with sane defaults
Most clients should not have to care about concepts such as "streaming", "buffering" and "parsing". In particular, most of our examples completely ignore this fact. As such, it makes more sense to add a new `Server` that takes care of all this for the 80% use case. We will likely rename our current `Server` implementation to something like `StreamingServer` and implement a simple facade that uses our `RequestBodyBufferMiddleware` and `RequestBodyParserMiddleware`.

Refs #197 and #198

### PR DESCRIPTION
**Pull request by WyriHaximus**
Follow up on #266 reintroducing the `Server` class but this time as a facade around `StreamingServer`.

- [x] Documentation
- [x] Tests

Closes #259

### CODE REVIEW

**Comment by WyriHaximus** [comment]
Ping @clue && @jsor the first bit of documentation is in.

**Comment by WyriHaximus** [comment]
Ping @clue new middleware is included in the Server facade now. Decided to go with a strategy that bases the amount of concurrently handled requests on the memory limit and post size limit.

File: src/Server.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+/**
+ * Facade around StreamingServer with sane defaults for 80% of the use cases.
+ * The requests passed to your callable are fully compatible with PSR-7 because
+ * the body of the requests are fully buffered and parsed, unlike StreamingServer
+ * where the body is a raw ReactPHP stream.
+ *
+ * Wraps StreamingServer with the following middleware:
+ * - RequestBodyBufferMiddleware
+ * - RequestBodyParserMiddleware
+ * - The callable you in passed as first constructor argument
```
Comment (jsor) [inline src/Server.php]: `LimitConcurrentRequestsMiddleware` missing now here.

Comment (WyriHaximus) [reply src/Server.php]: Fixed that :+1:

**Comment by WyriHaximus** [comment]
Added documentation to the readme

File: src/Server.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ *
+ * All middleware use their default configuration, which can be controlled with
+ * the the following configuration directives from php.ini:
+ * - upload_max_filesize
+ * - post_max_size
+ * - max_input_nesting_level
+ * - max_input_vars
```
Comment (clue) [inline src/Server.php]: This also respects `file_uploads` and `max_file_uploads` ini settings as documented for `RequestBodyParserMiddleware`.

File: src/Server.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+/**
+ * Facade around StreamingServer with sane defaults for 80% of the use cases.
+ * The requests passed to your callable are fully compatible with PSR-7 because
+ * the body of the requests are fully buffered and parsed, unlike StreamingServer
+ * where the body is a raw ReactPHP stream.
+ *
+ * Wraps StreamingServer with the following middleware:
+ * - LimitConcurrentRequestsMiddleware
+ * - RequestBodyBufferMiddleware
+ * - RequestBodyParserMiddleware
```
Comment (clue) [inline src/Server.php]: What about the `enable_post_data_reading` ini setting as documented for `RequestBodyParserMiddleware`? Arguably, we *may* want to remove the `RequestBodyParserMiddleware` from the middleware list in this case to mimic PHP's default behavior?

Comment (WyriHaximus) [reply src/Server.php]: Updated the PR to respect `enable_post_data_reading`

File: src/Server.php (hunk 2 above)
Comment (WyriHaximus) [reply src/Server.php]: Ah my bad, added them :+1:

File: src/Server.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    public function __construct($callback)
+    {
+        if (!is_callable($callback)) {
+            throw new \InvalidArgumentException();
+        }
+
+        $middleware = array();
+        $middleware[] = new LimitConcurrentRequestsMiddleware($this->getConcurrentRequestsLimit());
+        $middleware[] = new RequestBodyBufferMiddleware();
+        if (ini_get('enable_post_data_reading')) {
```
Comment (clue) [inline src/Server.php]: This fails with HHVM and legacy PHP < 5.4: https://3v4l.org/qJtsa

```
$ php -d enable_post_data_reading=no -r 'var_dump(ini_get("enable_post_data_reading"));'
string(0) ""
```

Accordingly, this should probably check for `!== ''` instead? Also, can you add a comment for this?

Comment (WyriHaximus) [reply src/Server.php]: Hmmm interesting, but what happens then when set to `0`?

Comment (clue) [reply src/Server.php]: According to http://php.net/manual/en/ini.core.php#ini.enable-post-data-reading it's a `boolean`.

Comment (WyriHaximus) [reply src/Server.php]: Updated the check

**Comment by WyriHaximus** [comment]
@clue fixed the test and squashed the commits

File: src/Server.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+ * - post_max_size
+ * - max_input_nesting_level
+ * - max_input_vars
+ * - file_uploads
+ * - max_file_uploads
+ * - enable_post_data_reading
+ */
+final class Server
+{
+    const MAXIMUM_CONCURRENT_REQUESTS = 100;
```
Comment (clue) [inline src/Server.php]: Minor nitpick: Should probably be `@internal` :+1:

Comment (WyriHaximus) [reply src/Server.php]: Good point, added that :+1:

**Pull request merged**